models/maillog.go

```diff
@@ -27,6 +32,10 @@ var MaxSendAttempts = 8
 // MailLog is exceeded.
 var ErrMaxSendAttempts = errors.New("max send attempts exceeded")
 
+// processAttachment is used to to keep track of which email attachments have templated values.
+// This allows us to skip re-templating attach
+var processAttachment = map[[20]byte]bool{} // Considered using attachmentLookup[campaignid][filehash] but given the low number of files current approach should be fine
```

Comment:
I'd prefer to err away from using a global map as a cache. While it wouldn't be immediately an issue, it takes up memory forever, not just when a campaign is being sent out.

Instead, just thinking out loud, but what we might be able to do is to modify the `models.Attachment` object to have an unexported `templateApplied` boolean field or something to indicate whether or not the templating has been applied to file already. This would be false by default then, after the `applyAttachmentTemplate` function has completed, we can set the flag so the attachment isn't processed again.

Additionally, I'd prefer to have `applyAttachmentTemplate` not be in `maillog.go`, but rather have it be a function off of an `Attachment` in `attachment.go`.

Ideally, it'd be something that kinda looks like this (I haven't tested this!)

in `models/attachment.go`:

```go
// Attachment contains the fields and methods for
// an email attachment
type Attachment struct {
	Id         int64  `json:"-"`
	TemplateId int64  `json:"-"`
	Content    string `json:"content"`
	Type       string `json:"type"`
	Name       string `json:"name"`

        templateApplied bool
}
...
func (a *Attachment) applyTemplate(ptx PhishingTemplateContext) (io.Reader, error) {
    if a.templateApplied {
        return strings.NewReader(a.Content), nil
    }
    // Do all your normal processing
    ...
    a.Content = processedAttachment
    a.templateApplied = true
    return strings.NewReader(a.Content), nil
}
```

Then, back in `maillog.go`, the call to `Attach` would look something like:

```go
msg.Attach(func(a *Attachment) (string, gomail.FileSetting, gomail.FileSetting) {
	h := map[string][]string{"Content-ID": {fmt.Sprintf("<%s>", a.Name)}}
	return a.Name, gomail.SetCopyFunc(func(w io.Writer) error {
		content := a.ApplyTemplate(ptx)
		_, err = io.Copy(w, content)
		return err
	}), gomail.SetHeader(h)
}(&a))
```

I _think_ something like this would work and would be cleaner, but give it a shot and let me know if you run into issues.

Reply:
Good thinking!

I'm not sure if the `templateApplied` boolean will work though, as we have three states for an attachment:

1. Never been checked to see if it has variables
2. Been checked, doesn't have variables
3. Been checked, does have variables

We could use an `int` and have 0, 1, 2 states to represent this. Or we could use a *bool as the default value would be null. I'll explore this now.

Reply:
Oh, thinking about this some more, we can just default to a state that yes "yes check this file" and change it to the inverse if no variables are found the first time processing it (which is actually the logic I use).

Reply:
I pushed code having a go at this approach and ran tests against my set of attachment files.

Reply:
Running some debug statements to make sure files that we know don't have template variables aren't being reprocessed, and I notice the code isn't working.

It seems the `a.templateApplied = true` (in my case a.noTemplateVars) logic doesn't 'stick', as on the next iteration it's false.

Perhaps I'm not understanding unexported variables correctly.

Reply:
I haven't tested, but I think the issue is that we're not passing a pointer to the attachment in the call to `msg.Attach`. Take a look at the code snippet above and you'll see that I changed it to be a pointer (e.g. `func(a *Attachment)...` then passing `(&a)` to that function.).

> Oh, thinking about this some more, we can just default to a state that says "yes check this file" and change it to the inverse if no variables are found the first time processing it (which is actually the logic I use). e.g:

Yep! That's generally what I was thinking. I could be wrong, but this feels like simply an optimization to avoid parsing the entire attachment for every email. That's why I was calling it `templateApplied`, since it was just a flag to say that we've already processed it for templating.

Reply:
Eh, thinking more about it, I should clarify the above comment. I was wrong to say "we only need to process the file once" since if it _does_ have template variables then we'll need to process it every time so that the right template variables are parsed, right?

But if there are no template variables, then you'd be right just to short-circuit and return the content as-is. Let me think a little bit about whether this would change our approach.

Reply:
Yeah if there are template variables we'll need to process the attachment every time.

So first pass we assume there are template variables, and at the end we set the flag to confirm/deny this.

Then on subsequent passes we either return immediately as there are no variables, or we parse the attachment for the variables.

Understood on passing the pointer now, yes that makes sense. I'll fiddle some more while you ponder to see if we're doing it right.

Reply:
I suspect I'm missing something. I've copy/pasted your `msg.Attach()` code (just added the error handling) with appropriate by-reference logic now:

```go
// Attach the files
fmt.Println("Handling attachments for user " + r.FormatAddress())
for _, a := range c.Template.Attachments {
    msg.Attach(func(a *Attachment) (string, gomail.FileSetting, gomail.FileSetting) {
        h := map[string][]string{"Content-ID": {fmt.Sprintf("<%s>", a.Name)}}
        return a.Name, gomail.SetCopyFunc(func(w io.Writer) error {
            content, err := a.ApplyTemplate(ptx)
            if err != nil {
                return err
            }
            _, err = io.Copy(w, content)
            return err
        }), gomail.SetHeader(h)
    }(&a))
}
```


Then in `attachment.go`:

```go
type Attachment struct {
	Id          int64  `json:"-"`
	TemplateId  int64  `json:"-"`
	Content     string `json:"content"`
	Type        string `json:"type"`
	Name        string `json:"name"`
	vanillaFile bool   // Vanilla file, when true, has no template variables
}
```

(vanillaFile denotes a file *without* variables)

and then the `ApplyTemplate` function:

```go
func (a *Attachment) ApplyTemplate(ptx PhishingTemplateContext) (io.Reader, error) {

	fmt.Printf("File ID: %d\nFile name: %s\nTemplate boolean: %t\n", a.Id, a.Name, a.vanillaFile)
	a.vanillaFile = true  // Hard set for testing

      ......
```

The boolean doesn't seem to propagate between instances of the attachment between different emails, even with the above 'hard set':
```
Handling attachments for user "Test 1" <test1@mail.com>
File ID: 90
File name: doc-without-template-vars.docx
Template boolean: false

Handling attachments for user "Test 2" <test2@mail.com>
File ID: 90
File name: doc-without-template-vars.docx
Template boolean: false

Handling attachments for user "Test 2" <test2@mail.com>
File ID: 90
File name: doc-without-template-vars.docx
Template boolean: false
```

If I'm not mistaken we're looping through multiple attachments on a single email with the above code, so if the email had the identical attachment multiple times this would work (not that that's possible in gophish). Once we move on to a *new* email and we reach this code again we don't have a reference to the boolean.

Reply:
Yes, that thread seemed to be the right one to pull - this seems to work now. Iterating over `c.Template.Attachments` by reference:

```go
for i, _ := range c.Template.Attachments {
	a := &c.Template.Attachments[i]
	msg.Attach(func(a *Attachment) (string, gomail.FileSetting, gomail.FileSetting) {
		h := map[string][]string{"Content-ID": {fmt.Sprintf("<%s>", a.Name)}}
		return a.Name, gomail.SetCopyFunc(func(w io.Writer) error {
			content, err := a.ApplyTemplate(ptx)
			if err != nil {
				return err
			}
			_, err = io.Copy(w, content)
			return err
		}), gomail.SetHeader(h)
	}(a))
}
```

This is probably what you meant in your original message and I read it wrong.